Inline ``store_chunk`` calls for ``store``'s ``return_stored`` option

jakirkham opened the pull request:
Inlines `store_chunk` calls inside `load_chunk` calls for `store` in an attempt to join tasks and reuse resources. This is handled slightly differently depending on whether the `store_chunk` tasks are run immediately or delayed.

In the first case where the `store_chunk` tasks are `persist`ed, the `Future`s are placed in the `load_chunk` calls and the graph is shrinked. Typically `distributed` inlines `Futures` when we submit the job amongst other steps. So doing this inlining in advance is fine. If this is run locally, then the results are inlined instead, which is still ok. This also halves the graph size while still containing all the same information.

In the second case where the `store_chunk` calls are delayed, the function is inlined into `load_chunk`. This should ensure that one task runs both functions one after another. Further the lock is held for the duration of both the `store_chunk` and `load_chunk` tasks. This should avoid some contention over the lock and allow running tasks with the lock to complete without getting blocked by other tasks that could have acquired the lock between `store_chunk` and `load_chunk` tasks previously.

More generally this should simplify the Dask graph a bit and improve performance. Given the effect of this inlining on the graph (and existing nomenclature in the code), we rename `load_chunk` to `load_store_chunk`.

Also improves test coverage of the different options for returning results from `store`.

jakirkham commented on dask/array/core.py:
```diff
@@ -930,12 +937,23 @@ def store(sources, targets, lock=True, regions=None, compute=True,
+            from ..optimization import cull, inline
+            load_dsks_mrg = cull(inline(load_dsks_mrg), load_keys)[0]
+        else:
+            from ..optimization import inline_functions
```
Trying to import either of these outside the function results in an `ImportError` during testing. So have imported them locally only where they are needed to avoid this issue.

jakirkham commented on dask/array/core.py:
```diff
@@ -2657,14 +2677,19 @@ def store_chunk(x, out, index, lock, region, return_stored):
+    except:
         if lock:
             lock.release()
+        raise
```
Need to make sure we always release the lock if there was an exception as we won't make it to `load_store_chunk` to release it.

jakirkham commented on dask/array/core.py:
```diff
@@ -2657,14 +2679,19 @@ def store_chunk(x, out, index, lock, region, return_stored):
-    finally:
+    except:  # noqa: E722
         if lock:
             lock.release()
```
Need to make sure we always release the lock if there was an exception as we won't make it to `load_store_chunk` to release it.

jakirkham replied:
Need to make sure we always release the lock if there was an exception as we won't make it to `load_store_chunk` to release it.

jakirkham commented on dask/array/core.py:
```diff
@@ -2657,14 +2679,19 @@ def store_chunk(x, out, index, lock, region, return_stored):
+    except:  # noqa: E722
         if lock:
             lock.release()
+        raise
```
Need to make sure we always release the lock if there was an exception as we won't make it to `load_store_chunk` to release it.

jakirkham commented on dask/array/core.py:
```diff
@@ -929,13 +936,21 @@ def store(sources, targets, lock=True, regions=None, compute=True,
         store_dsks, tgt_dsks, [sources_dsk]
     ))
 
+    from ..optimization import cull
```
Trying to import this outside the function results in an `ImportError` during testing. So imported it locally only where it is needed to avoid this issue.

jakirkham commented on dask/array/core.py:
```diff
@@ -929,13 +936,21 @@ def store(sources, targets, lock=True, regions=None, compute=True,
-        load_dsks_mrg = sharedict.merge(store_dsks_mrg, *load_dsks)
+        load_dsks_mrg = sharedict.merge(load_dsks_mrg, store_dsks_mrg)
+
+        from ..optimization import inline
```
Trying to import this outside the function results in an `ImportError` during testing. So imported it locally only where it is needed to avoid this issue.

jakirkham commented on dask/array/core.py:
```diff
@@ -937,6 +941,9 @@ def store(sources, targets, lock=True, regions=None, compute=True,
 
         load_dsks_mrg = sharedict.merge(store_dsks_mrg, *load_dsks)
 
+        from ..optimization import cull, inline
```
Trying to import these outside the function results in an `ImportError` during testing. So imported them locally only where they are needed to avoid this issue.

mrocklin replied:
Do we know what the import issue is?  I would expect the `optimization` module to be pretty well isolated.

jakirkham replied:
Not really sure. Initially this issue forced me to keep it in the function. Now I'm having trouble reproducing it. 😕

jakirkham replied:
Have moved the `import` out now and it seems to work fine. Not sure what was going on during development.

jakirkham commented:
Outside of the `import` issues, any thoughts on these changes?

jakirkham commented:
Planning on merging tomorrow if no comments.

jakirkham commented:
Not to push, but have you had any time to look at this? If not, that's ok. Though do you know roughly when you might? Have been playing a bit with these changes locally and they seem ok. Am curious to know about what issues you perceive and how we can go about addressing them.

jcrist commented:
Apologies for the delay here, it's been a busy week. I expect to get to this first thing tomorrow morning.

jcrist commented:
Apologies for the delay, finally had time to review this.

My issue with this PR is it seems a bit complicated for the task it's trying to do. If I understand correctly, there are 4 cases we need to handle here:

1. `return_stored=False` and `compute=False`

Generate a graph with one `store_chunk` call per array chunk and return a `Delayed` containing that graph.

2. `return_stored=False and compute=True`

Generate a graph with one `store_chunk` call per array chunk, and call `compute` on the resulting graph.

3. `return_stored=True and compute=False`

Generate a graph with one `load_store_chunk` call per array chunk. This function takes the array chunk, stores it, and returns a new chunk reloaded from the store. The lock (if any) is held and released during whole store/load operation, entirely within the same function. Return a new dask array containing the resulting graph.

4. `return_stored=True and compute=True`

Generate a graph with one `store_chunk` call per array chunk. Compute this graph. Then form a new graph with one `load_chunk` call per array chunk. Return a new dask array containing the resulting graph.

The optimizations in this PR deal with 3. and 4. alone. It seems to me that it'd be easier to directly generate the graphs as I've described above than to try and use the optimization passes and some tricky locking behavior. The optimization passes aren't free, and the scheduler/collections also call them, potentially duplicating work.

Also, I'm not comfortable with functions that acquire and don't release locks, this seems prone for error. Exceptions don't bubble through tasks the same way that they do in python functions:

```python
def get_and_release(lock):
    with lock:
        do_stuff()

def get(lock):
    lock.acquire()
    do_stuff()
    return lock

def release(lock):
    lock.release()

{'safe': (get_and_release, lock),
 'dangerous': (release, (get, lock))}
```

In the `'safe'` case, the lock is always released if an exception is raised, whether from in the task or external in the scheduler (or even from a signal like `KeyboardInterrupt`). In the `'dangerous'` case, exceptions can stop the task processing without ever releasing the lock. The dask scheduler basically does:

```
temp = get(lock)
# What if an exception happened here?
dangerous = release(temp)
```

This might work fine 99% of the time, but I'm not comfortable with locking behavior like this in dask, as I can see future maintenance being potentially tricky.

---

I think if you broke this code up into three task functions: `store_chunk`, `load_chunk`, and `load_store_chunk`, and were careful about the way graphs were generated you wouldn't need the tricky lock behavior or the optimization passes here.

jakirkham commented:
Thanks for taking a look. To summarize, your concerns are the locking between functions is fragile and we shouldn't be performing optimization passes, correct?

jcrist commented:
Correct. Sometimes calling optimization passes in the graph building code is ok, but if possible we'd like to avoid it. In this case I think it'd be possible to generate equivalent graphs in a reasonable way without calling the optimizations.

jakirkham commented:
Was trying to handle exceptions that might mess up releasing the lock by adding the bare `except` in `store_chunk`. Are there other places where you can imagine an exception messing it up?

As far as overhead from optimizations, I generally agree with this, but felt the finality of `store` made it unique in this regard. Also given that we recently added optimizations to the input arrays in `store`, figured this was fair game and possibly would incur less of a penalty. Are there particular optimizations here that appear problematic?

Am asking these questions just as much about this context as trying to generally better understand what dask/distributed is doing under the hood in these cases.

jcrist commented:
> Was trying to handle exceptions that might mess up releasing the lock by adding the bare except in store_chunk. Are there other places where you can imagine an exception messing it up?

If an exception happens in the scheduler, or in anywhere outside the task function (perhaps due to a signal like `KeyboardInterrupt`. In general, I am -1 on any function that acquires a lock and doesn't also release it. Look at the "dangerous" example I provided above, the scheduler will run the `get` and `release` functions separately, if an exception occurs outside the task after `get` but before `release` the lock will never be released.

> As far as overhead from optimizations, I generally agree with this, but felt the finality of store made it unique in this regard. Also given that we recently added optimizations to the input arrays in store, figured this was fair game and possibly would incur less of a penalty. Are there particular optimizations here that appear problematic?

Nothing particular, but it is repeating work (most scale O(N), but not all). I feel this PR can be done cleaner without using the optimization passes. Optimizations should be used sparingly IMO, in favor of explicitly creating the graph as originally intended - they add overhead, can be fragile to change, and don't always make the intent (in this case load and store being in the same task) clear.

jakirkham commented:
So have an idea as to how to fix this. That said, think this needs a total rewrite to complete address the issue. Going to close this out and put a smaller PR together that takes a few minor optimizations from this one for now. Will then follow-up with a PR for some of the larger ideas in this one.

jakirkham commented:
Split out the tests from here into PR ( https://github.com/dask/dask/pull/3152 ). Also split out some minor optimizations from here in PR ( https://github.com/dask/dask/pull/3153 ). Will follow-up on the optimizations for `return_stored=True` and `compute=False` in a subsequent PR that resolves the locking issues.
